### ISSUE TITLE
Add -Wimplicit-fallthrough=0 to Makefile ?

### ISSUE THREAD

**Issue by matrixise** [python/cpython#75106]
BPO | [30923](https://bugs.python.org/issue30923)
--- | :---
Nosy | @vstinner, @skrah, @zware, @serhiy-storchaka, @matrixise, @stratakis
PRs | <li>python/cpython#2698</li><li>python/cpython#3132</li><li>python/cpython#3142</li><li>python/cpython#3157</li><li>python/cpython#3205</li><li>python/cpython#3518</li><li>python/cpython#4620</li>
Files | <li>[output.txt](https://bugs.python.org/file47012/output.txt "Uploaded as text/plain at 2017-07-13.16:35:58 by @matrixise")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-08-25.12:23:13.249>
created_at = <Date 2017-07-13.16:36:00.767>
labels = ['build', '3.7']
title = 'Add -Wimplicit-fallthrough=0 to Makefile ?'
updated_at = <Date 2017-11-29.23:00:37.850>
user = 'https://github.com/matrixise'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-29.23:00:37.850>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-08-25.12:23:13.249>
closer = 'skrah'
components = []
creation = <Date 2017-07-13.16:36:00.767>
creator = 'matrixise'
dependencies = []
files = ['47012']
hgrepos = []
issue_num = 30923
keywords = []
message_count = 27.0
messages = ['298299', '298300', '298302', '298306', '298307', '298312', '299063', '299070', '299072', '299085', '299086', '299089', '299090', '299091', '299092', '299093', '300483', '300490', '300526', '300527', '300585', '300620', '300622', '300833', '300835', '302003', '307269']
nosy_count = 6.0
nosy_names = ['vstinner', 'skrah', 'zach.ware', 'serhiy.storchaka', 'matrixise', 'cstratak']
pr_nums = ['2698', '3132', '3142', '3157', '3205', '3518', '4620']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'compile error'
url = 'https://bugs.python.org/issue30923'
versions = ['Python 3.7']
```

</p></details>

**Comment by matrixise** [comment]
Hi all,

Since I use the last version of Fedora 26 with gcc-7.1.1, I have these warnings (see output.txt file)

We could add -Wimplicit-fallthrough=0 to Makefile ? it will disable the fallthrough of the coed.

What do you think about that ? What's your feedback on this option and can we use it in the case of CPython ?

Thank you

**Comment by serhiy-storchaka** [comment]
It seems to me that by default the compiler recognizes a wide variety of "falls through" comments. Thus we need just add missed comments.

There are warnings emitted when compile imported third-party code. We should ask Stefan for adding "falls through" comments in his libmpdec. And either compile expat with -Wimplicit-fallthrough=0, or wait until warnings will be fixed in upstream, or remove the bundled copy of expat.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
It's a useful warning, but I find it annoying to add 20 "fall through" comments. I may add a pragma at some point.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
New changeset 72b543308ee3087e3fa247981f5cb4be1138c515 by Stefan Krah in branch 'master':
bpo-30923: Suppress fall-through warnings in libmpdec. (bpo-2698)
https://github.com/python/cpython/commit/72b543308ee3087e3fa247981f5cb4be1138c515

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Hmm, that took about 20 min to commit a 3 line diff.  Now I'm watching the buildbots...

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Stéphane, if you want the libmpdec change cherry picked and are willing to do the (significant) work of backporting, I'll merge it.

**Comment by matrixise** [comment]
@stefan

In fact I could create the PR for the backports, but I have again the same issues for the rest of CPython.

I have checked your code, and in my case 'gcc' is in the cc string, 'gcc -pthread' and you have fixed for libmpdec.

But for the other modules ?

**Comment by vstinner** [comment]
Hum, I don't think that it's worth it to backport changes which only fix warnings. Usually, we first focus on fixing all warnings on master.

**Comment by matrixise** [comment]
yep, currently, 3.6 and 3.5 are in 'bug fix' mode, and in this case, it's not a bug.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Well, it's not a bug, but perhaps it is annoying for users of this gcc version if they compile 3.6 often.


Actually, I think gcc should not include this warning in -Wextra.  It's something that could be run manually before a release.


I'd vote for making -Wno-implicit-fallthrough global.

**Comment by matrixise** [comment]
Stefan, ask on python-dev ml, or we have to ask to Ned Deily for this version ?

**Comment by vstinner** [comment]
I'm ok to kill warnings, but I would suggest to first fix most GCC7 warnings because starting to discuss backports. I expect that it will require multiple changes.

**Comment by matrixise** [comment]
I am not for a backport, it's not a security fix or a bug fix.

for my case, I just want to "kill" the warnings, maybe we could check the version of gcc and add "-Wimplicit-fallthrough=0".

Here is a good explanation of the 'fallthrough' in gcc7 
https://developers.redhat.com/blog/2017/03/10/wimplicit-fallthrough-in-gcc-7/

**Comment by 12c7b821-ad6f-4235-91d8-57b2aacb2dfc** [comment]
Yeah the warnings are quit annoying when compiling the master and 3.6 branch.

Fedora 26 currently includes gcc 7 which emits those warnings, other distros will follow up when they update gcc, so I'd think it would be better to fix it.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
We can check for the version, but all versions of gcc that I tested
accept and ignore -Wno-implicit-fallthrough, even though they don't 
actually have -Wimplicit-fallthrough.

Of course they choke on -Wimplicit-fallthrough=0.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
I think the fall-through blog notes are slightly overstated. :-)

"The switch fallthrough has been widely considered a design defect in C."


It's an important feature.

**Comment by vstinner** [comment]
The following commit broke \_decimal compilation on the "x86 Tiger 3.x" buildbot:

commit 72b543308ee3087e3fa247981f5cb4be1138c515
Author: Stefan Krah \<skrah@bytereef.org\>
Date:   Thu Jul 13 20:54:20 2017 +0200

    bpo-30923: Suppress fall-through warnings in libmpdec. (bpo-2698)

http://buildbot.python.org/all/builders/x86%20Tiger%203.x/builds/1068/steps/compile/logs/stdio

building '\_decimal' extension
gcc -fno-strict-aliasing -Wsign-compare -g -O0 -Wall -Wstrict-prototypes -std=c99 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -DUNIVERSAL=1 -I/Users/db3l/buildarea/3.x.bolen-tiger/build/Modules/_decimal/libmpdec -I./Include -I. -I/usr/local/include -I/Users/db3l/buildarea/3.x.bolen-tiger/build/Include -I/Users/db3l/buildarea/3.x.bolen-tiger/build -c /Users/db3l/buildarea/3.x.bolen-tiger/build/Modules/_decimal/_decimal.c -o build/temp.macosx-10.4-i386-3.7-pydebug/Users/db3l/buildarea/3.x.bolen-tiger/build/Modules/_decimal/_decimal.o -Wno-implicit-fallthrough
cc1: error: unrecognized command line option "-Wno-implicit-fallthrough"

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Thanks. I tried to revert it, but got:

remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
remote: error: [GH006](https://github.com/python/cpython/pull/006): Protected branch update failed for refs/heads/master.
remote: error: 2 of 2 required status checks are expected.
To https://github.com/python/cpython.git
 ! [remote rejected]       master -\> master (protected branch hook declined)
error: failed to push some refs to 'https://github.com/python/cpython.git'


So I'm supposed to spend 20 min creating an irrelevant PR to revert a 3 line diff?!

I would never have finished \_decimal under these conditions.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
So I installed gcc-7.2.0 from source. Hilariously compiling gcc \*itself*
emits fallthrough warnings!

Then I tried to be a good open source drone and add 20 /* fall through \*/
comments to libmpdec.

gcc is too stupid to recognize the /* fall through \*/ at the #endif
position.


Perhaps the author of this gcc warning wants to submit a patch.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
New changeset d73a960c575207539c3f9765cff26d4fff400b45 by Stefan Krah in branch 'master':
bpo-30923: Disable warning that has been part of -Wextra since gcc-7.0. (bpo-3142)
https://github.com/python/cpython/commit/d73a960c575207539c3f9765cff26d4fff400b45

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
[PR 3157](https://github.com/python/cpython/pull/3157) addresses everything apart from expat and

  https://github.com/python/cpython/blob/master/Modules/cjkcodecs/_codecs_iso2022.c#L816


I'm not sure about that one. It looks harmless but a bit odd.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
New changeset f432a3234f9f2ee09bd40be03e06bf72865ee375 by Stefan Krah in branch 'master':
bpo-30923: Silence fall-through warnings included in -Wextra since gcc-7.0. (bpo-3157)
https://github.com/python/cpython/commit/f432a3234f9f2ee09bd40be03e06bf72865ee375

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Cherry picking has too many conflicts, I'm not backporting this myself.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
New changeset 9e1e6f528f3fec16b9bd99f5ee38048ffec04a81 by Stefan Krah in branch 'master':
bpo-30923: Silence fall-through warnings in libexpat build. (bpo-3205)
https://github.com/python/cpython/commit/9e1e6f528f3fec16b9bd99f5ee38048ffec04a81

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
All warnings except for bpo-31275 are dealt with.

**Comment by vstinner** [comment]
New changeset c0e77364ca29df6cfb311e79892955c92bd8e595 by Victor Stinner in branch '3.6':
[3.6] bpo-30923: Silence fall-through warnings included in -Wextra since gcc-7.0 (bpo-3518)
https://github.com/python/cpython/commit/c0e77364ca29df6cfb311e79892955c92bd8e595

**Comment by vstinner** [comment]
New changeset dedcbee04cd52790027ecfb46cb3aa33efebdc84 by Victor Stinner in branch '3.6':
[3.6] bpo-30923, bpo-31279: Fix GCC warnings (bpo-4620)
https://github.com/python/cpython/commit/dedcbee04cd52790027ecfb46cb3aa33efebdc84
